exportable case reducers type, bugfix, documentation

phryneas opened the pull request:
This is what I'm currently working on and it has already gotten a bit bigger than expected :smile: 

INCLUDED:
- [x] change signatue of `createSlice` and add new types (exported) to be used in situations like #276 and #286 :
  - `SliceCaseReducerDefinitions`
  - `CaseReducerActions`
  - `CaseReducerWithPrepare` (not really necessary to export this, but `CaseReducer` was already exported so it might not hurt to have this)
- [x] fix a bug where `PayloadActionCreator<any>` becomes `ActionCreatorWithoutPayload` which was a regression introduced in #273 (also opened https://github.com/joonhocho/tsdef/pull/4 )
- [x] type `_ActionCreatorWithPreparedPayload` is now marked `@internal` (was `ActionCreatorWithPreparedPayload` before #273, since we renamed it anyway might as well hide it completely)
- [x] `@internal` types will be stripped from the `dist/typings.d.ts` file
- [x] document ALL types
- [x] add release tags (@public or @internal) to all types
- [x] add a type test that successfully wraps `createSlice`
- [x] ~~api-extractor trims too agressively (complete removal instead of "not export") - I'll have to see if that works out. (see https://github.com/microsoft/rushstack/issues/1664 )~~ (solved by manually exporting)
- [x] fix TS 3.3
- [x] add "Wrapping createSlice" chapter to the TS docs 
- [x] add "Defining the type of your initialState"  chapter to the TS docs

markerikson commented on src/createSlice.ts:
```diff
@@ -84,84 +94,97 @@ export interface CreateSliceOptions<
-}
+/**
+ * The type describing a slice's `reducers` option.
+ * Also checks itself, so it has to be passed "itself" as it's second option.
```
wat :)

phryneas replied:
Yeah, I wouldn't have found out about that one myself as it is completely counter-intuitive. But a generic can `extend` (which essentially just means "match this shape") something that *depends on the generic being matched*.

So you can write
```ts
function something<X extends Restrict<X>>(x: X) {}
```

and then `Restrict` can be a type that does some crazy restrictions on the original input object while referencing it. 

This is something I learned from @jcalz over on StackOverflow: https://stackoverflow.com/questions/59200095/is-there-a-simpler-way-of-associating-a-key-to-a-value-in-a-mapped-obect?noredirect=1#comment104621839_59200095
Thank you again Joe!

markerikson replied:
Still waiting for that blog post / talk on "all the bizarre stuff I learned working on RTK's TS types" :)

phryneas replied:
Yup, got something in the making :) Meetup somewhere mid-february, maybe I'll also submit it to a few conferences. Also, some ideas to experiment with twitter-content :)

markerikson commented:
Your ability to crank out TS improvements continues to amaze me :)

No specific feedback from me on the changes, but as usual I still don't know quite enough to provide much feedback here anyway.  

Do you want to try to get the other type doc changes in here too, or save those for later?

phryneas commented:
> Your ability to crank out TS improvements continues to amaze me :)

This project is a wonderful learning experience, but it kills time like nothing else. I think i've started this Refactor over four or five times, each time putting in like 3-5 hours and then throwing everything away because it just wasn't working out.
So in the end, it just comes down to trial, error and my determination to bash my head against it until I find out how it works ;)

> Do you want to try to get the other type doc changes in here too, or save those for later?

I'm planning on doing everything in this PR, now that I already mixed different things in here. 
We'll probably have to wait for https://github.com/microsoft/rushstack/issues/1664 anyways and I need to find out why this doesn't work with TS3.3 again :woman_shrugging:

phryneas commented:
So it seems that relying on `@internal` won't be a possibility for us to exclude certain types from the dts-rollup (although it has a nice documenting effect when reading the source :smile: ). Which means I'll have to replace the `export * from ...` in `index.ts` with explicit export statements.
@markerikson would that be okay with you? I believe that might even help bundlers to do tree-shaking, although that's only a gut feeling.

The alternative would be to move the type to a different file where nothing is exported from, but right now I'd say that the type is in the file it belongs to.

markerikson commented:
Yeah, if that's necessary to keep internal types from leaking, we can do that.

phryneas commented:
Okay, that's it for today. All that's left is fixing TS3.3 - I hope that's gonna work out, but I'll leave that problem for "future me" :smile:

phryneas commented:
So, to make this work with TS <3.4, unfortunately I had to move `ValidateSliceCaseReducers` out of `SliceCaseReducers` (I renamed that one from `SliceCaseReducerDefintions`) again, because the `generic that extends itself to restrict itself` tactic seems not to be working in older TS versions.
This means, we have one more exported type: `ValidateSliceCaseReducers`.

But all in all, given that `SliceCaseReducers` is now a normal type and we aren't going over that "anoymous `PayloadActions` type" that was in there that was absolutely confusing, I think this whole refactor still has some good value and we can actually expose those types in good consciousness.

phryneas commented on type-tests/files/createSlice.typetest.ts:
```diff
@@ -312,3 +314,57 @@ function expectType<T>(t: T) {
+        ...reducers
+      }
+    })
+  }
```
This is what wrapping `createSlice` would look like now.

phryneas commented:
Oh, seriously? Now TS is acting up, only in 3.4 :roll_eyes: 

`Draft<GenericState<T>>` is not compatible with `GenericState<T>`. Suure.

phryneas commented:
Yay, we're all green, without dirty hacks. 

I was almost rage-quitting there o_O

markerikson commented on src/createReducer.ts:
```diff
@@ -7,6 +7,11 @@ import {
  * Defines a mapping from action types to corresponding action object shapes.
+ *
+ * @deprecated This should not be used manually - it is only used for internal
+ *             interference purposes and should not have any further value.
```
Should this be `inference` instead of `interference`?

phryneas replied:
Yup. Thanks for catching!

markerikson commented on src/createReducer.ts:
```diff
@@ -31,6 +38,12 @@ export type CaseReducer<S = any, A extends Action = AnyAction> = (
  * A mapping from action types to case reducers for `createReducer()`.
+ *
+ * @deprecated This should not be used manually - it is only used
+ *             for internal interference purposes and using it manually
```
Ditto.

markerikson commented on src/createSlice.ts:
```diff
@@ -42,18 +47,20 @@ export interface Slice<
 
-  caseReducers: SliceDefinedCaseReducers<CaseReducers, State>
+  /**
+   * The reducers defined by `reducers` for easy access if they were defined inline when calling createSlice.
```
```suggestion
   * The individual case reducer functions that were passed in the `reducers` parameter.  
   * This enables reuse and testing if they were defined inline when calling `createSlice`.
```

markerikson commented on type-tests/files/createSlice.typetest.ts:
```diff
@@ -312,3 +314,57 @@ function expectType<T>(t: T) {
   })
 }
+
+/** Test: wrapping createSlice should be possible */
```
Nice to see that A) this is possible, and B) we have a test for it.

markerikson replied:
Also, can we add this example to the "Usage with TS" docs page?

phryneas replied:
Honestly, without these tests I wouldn't have the confidence to change on single letter in the typings. Those are so invaluable!

markerikson commented:
No immediate complaints about the TS stuff, but my eyes also mostly glazed over, so take it as you will :)

Saw a couple possible comment typos, suggested a rewording, and it'd be nice to see the "generic slice" example in the docs, but other than that looks okay.

phryneas commented:
Added a "Wrapping `createSlice`" chapter to the TS docs. While I was at it, I also added a "Defining the type of your `initialState`" chapter, as that was something of importance we had not mentioned before.

markerikson commented:
Awright, let's get this in.
